lib/mix/tasks/compile.thrift.ex

```diff
@@ -71,7 +71,12 @@ defmodule Mix.Tasks.Compile.Thrift do
   end
 
   defp generate(thrift_file, output_dir, options) do
-    args = ~w[--gen erl --out] ++ [output_dir] ++ options ++ [thrift_file]
+    # if --gen is in the options given, leave it out from the base args
+    baseargs = case Enum.member?(options, "--gen") do
+      true  -> ~w[--out] ++ [output_dir]
+      false -> ~w[--gen erl --out] ++ [output_dir]
+    end
+    args = baseargs ++ options ++ [thrift_file]
```

Comment:
`generate` is called once for each Thrift file. It would be better to do this work earlier (in `run`) and pass a full set of options to generate. A `get_thrift_options` private helper routine would make the code pretty clear.

`generate` would then just append the filename and invoke the `thrift` binary.